packages/alfa-rules/src/sia-r15/rule.ts

```diff
@@ -38,37 +40,39 @@ export default Rule.Atomic.of<Page, Iterable<Element>, Question>({
             )
           );
 
-        const roots = iframes.groupBy((iframe) => iframe.root());
-
-        return flatMap(roots.values(), (iframes) =>
-          iframes
-            .reduce((groups, iframe) => {
-              for (const [node] of Node.from(iframe, device)) {
-                const name = node.name.map((name) => name.value);
-
-                groups = groups.set(
-                  name,
-                  groups
-                    .get(name)
-                    .getOrElse(() => List.empty<Element>())
-                    .append(iframe)
-                );
-              }
-
-              return groups;
-            }, Map.empty<Option<string>, List<Element>>())
-            .values()
-        );
+        const groups = iframes
+          .reduce((groups, iframe) => {
+            for (const [node] of Node.from(iframe, device)) {
+              const name = node.name.map((name) => name.value);
+
+              groups = groups.set(
+                name,
+                groups
+                  .get(name)
+                  .getOrElse(() => List.empty<Element>())
+                  .append(iframe)
+              );
+            }
+
+            return groups;
+          }, Map.empty<Option<string>, List<Element>>())
+          .values();
+        return filter(groups, (group) => group.size > 1);
       },
 
       expectations(target) {
-        const sources = Set.from(
-          map(target, (iframe) => iframe.attribute("src"))
+        const source = commonValue(
+          List.from(target).map((iframe) =>
+            embeddedResource(iframe, response.url)
+          )
         );
 
         return {
           1: expectation(
-            sources.size === 1,
+            source.isSome() &&
+              // always ask question if we can't find source, presumably the content doc is fed another way into the iframes
+              source.get() !== "invalid:" &&
+              source.get() !== "nothing:",
```

Comment:
Seems like these cases should simply be `None`.

Reply:
🤔 that would break the agnosticity of `commonValue`.
OTOH, `embeddedResource` should have returned a `Result` from the start and that makes this test much nicer indeed 👍